File: placement-prep/apps/api/src/lib/qdrant.ts
import { QdrantClient } from '@qdrant/js-client-rest';

const QDRANT_URL = process.env.QDRANT_URL || 'http://localhost:6333';
const QDRANT_API_KEY = process.env.QDRANT_API_KEY || '';
const COLLECTION_NAME = process.env.QDRANT_COLLECTION_NAME || 'leetcode_problems';

export const qdrantClient = new QdrantClient({
  url: QDRANT_URL,
  apiKey: QDRANT_API_KEY,
});

/**
 * Matches the ACTUAL payload stored in the Qdrant `leetcode_problems` collection.
 * These fields were set during ingestion in `ingest_to_qdrant_local.ipynb`.
 *
 * NOTE: `company`, `frequency`, and `acceptance` data lives in
 * Supabase `lc_company_questions` — NOT in Qdrant.
 */
export interface QdrantQuestionPayload {
  id: number;
  title: string;
  slug: string;
  difficulty: string; // "Easy" | "Medium" | "Hard"
  category: string;   // "Algorithms" | "Database" | "JavaScript" | etc.
  topic_tags: string[];
}

/**
 * Scroll Qdrant collection with payload-level filters.
 * No vector search — pure metadata filtering.
 */
export async function scrollQuestions(params: {
  difficulty?: string;
  topicTags?: string[];
  limit?: number;
}): Promise<QdrantQuestionPayload[]> {
  const { difficulty, topicTags, limit = 100 } = params;

  const mustFilters: any[] = [];

  if (difficulty) {
    mustFilters.push({
      key: 'difficulty',
      match: { value: difficulty },
    });
  }

  if (topicTags && topicTags.length > 0) {
    // Each tag is an "any" match — at least one of the provided tags must be present
    mustFilters.push({
      key: 'topic_tags',
      match: { any: topicTags },
    });
  }

  try {
    const results = await qdrantClient.scroll(COLLECTION_NAME, {
      filter: mustFilters.length > 0 ? { must: mustFilters } : undefined,
      limit,
      with_payload: true,
      with_vector: false,
    });

    const questions: QdrantQuestionPayload[] = [];

    if (results.points) {
      for (const point of results.points) {
        const payload = point.payload as Record<string, any>;
        questions.push({
          id: payload.id ?? (typeof point.id === 'number' ? point.id : 0),
          title: payload.title ?? '',
          slug: payload.slug ?? '',
          difficulty: payload.difficulty ?? 'Medium',
          category: payload.category ?? 'Algorithms',
          topic_tags: payload.topic_tags ?? [],
        });
      }
    }

    return questions;
  } catch (error) {
    console.error('Qdrant scroll error:', error);
    return [];
  }
}

/**
 * Perform semantic (vector) search against the collection.
 * Requires a pre-computed query vector (384-dim for bge-small-en-v1.5).
 */
export async function semanticSearch(params: {
  queryVector: number[];
  difficulty?: string;
  topicTags?: string[];
  limit?: number;
}): Promise<QdrantQuestionPayload[]> {
  const { queryVector, difficulty, topicTags, limit = 20 } = params;

  const mustFilters: any[] = [];

  if (difficulty) {
    mustFilters.push({ key: 'difficulty', match: { value: difficulty } });
  }
  if (topicTags && topicTags.length > 0) {
    mustFilters.push({ key: 'topic_tags', match: { any: topicTags } });
  }

  try {
    const results = await qdrantClient.search(COLLECTION_NAME, {
      vector: queryVector,
      filter: mustFilters.length > 0 ? { must: mustFilters } : undefined,
      limit,
      with_payload: true,
    });

    return results.map((point) => {
      const payload = point.payload as Record<string, any>;
      return {
        id: payload.id ?? 0,
        title: payload.title ?? '',
        slug: payload.slug ?? '',
        difficulty: payload.difficulty ?? 'Medium',
        category: payload.category ?? 'Algorithms',
        topic_tags: payload.topic_tags ?? [],
      };
    });
  } catch (error) {
    console.error('Qdrant semantic search error:', error);
    return [];
  }
}

File: placement-prep/apps/ai-service/app/core/qdrant.py
from qdrant_client import QdrantClient
from qdrant_client.models import Filter, FieldCondition, MatchValue, MatchAny
from app.config import settings

_client = None

def get_client() -> QdrantClient:
    global _client
    if _client is None:
        _client = QdrantClient(url=settings.QDRANT_URL, api_key=settings.QDRANT_API_KEY)
    return _client

async def semantic_search(
    query_vector: list[float],
    topic_filters: list[str] | None = None,
    difficulty: str | None = None,
    limit: int = 20
) -> list[dict]:
    client = get_client()
    
    must_conditions = []
    
    if difficulty:
        must_conditions.append(
            FieldCondition(key="difficulty", match=MatchValue(value=difficulty))
        )
    
    if topic_filters and len(topic_filters) > 0:
        must_conditions.append(
            FieldCondition(key="topic_tags", match=MatchAny(any=topic_filters))
        )
        
    query_filter = Filter(must=must_conditions) if must_conditions else None
    
    response = client.query_points(
        collection_name=settings.QDRANT_COLLECTION_NAME,
        query=query_vector,
        query_filter=query_filter,
        limit=limit,
        with_payload=True
    )
    results = response.points
    
    return [
        {
            "id": point.payload.get("id"),
            "title": point.payload.get("title"),
            "slug": point.payload.get("slug"),
            "difficulty": point.payload.get("difficulty"),
            "category": point.payload.get("category"),
            "topic_tags": point.payload.get("topic_tags", []),
            "score": point.score
        }
        for point in results
    ]

async def scroll_by_tags(
    topic_filters: list[str] | None = None,
    difficulty: str | None = None,
    limit: int = 50
) -> list[dict]:
    client = get_client()
    
    must_conditions = []
    
    if difficulty:
        must_conditions.append(
            FieldCondition(key="difficulty", match=MatchValue(value=difficulty))
        )
    
    if topic_filters and len(topic_filters) > 0:
        must_conditions.append(
            FieldCondition(key="topic_tags", match=MatchAny(any=topic_filters))
        )
        
    query_filter = Filter(must=must_conditions) if must_conditions else None
    
    results, _next_page = client.scroll(
        collection_name=settings.QDRANT_COLLECTION_NAME,
        scroll_filter=query_filter,
        limit=limit,
        with_payload=True,
        with_vectors=False
    )
    
    return [
        {
            "id": point.payload.get("id"),
            "title": point.payload.get("title"),
            "slug": point.payload.get("slug"),
            "difficulty": point.payload.get("difficulty"),
            "category": point.payload.get("category"),
            "topic_tags": point.payload.get("topic_tags", [])
        }
        for point in results
    ]
